src/data/blob.rs

```diff
@@ -0,0 +1,420 @@
+    fn serialize<S: Serializer>(&self, serialiser: S) -> Result<S::Ok, S::Error> {
+        self.value.serialize(serialiser)
+    }
+}
+
+impl<'de> Deserialize<'de> for PublicBlob {
+    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        let value: Vec<u8> = Deserialize::deserialize(deserializer)?;
+        Ok(PublicBlob::new(value))
+    }
+}
+
+impl Debug for PublicBlob {
+    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
+        write!(formatter, "PublicBlob {:?}", self.name())
+    }
+}
+
+/// The Kind determines if a blob is public or private.
+#[derive(Clone, Copy, Eq, PartialEq, Ord, PartialOrd, Hash, Serialize, Deserialize, Debug)]
+pub enum Kind {
+    /// Denotes a blob as private.
+    Private,
+    /// Denotes a blob as public.
+    Public,
+}
+
+impl Kind {
+    /// Returns true if public.
+    pub fn is_public(self) -> bool {
+        self == Kind::Public
+    }
+
+    /// Returns true if private.
+    pub fn is_private(self) -> bool {
+        !self.is_public()
+    }
+
+    /// Creates `Kind` from a `public` flag.
+    pub fn from_flag(public: bool) -> Self {
+        if public {
+            Kind::Public
+        } else {
+            Kind::Private
+        }
+    }
+}
+
+/// A blob address, where public and private are two different address spaces.
+#[derive(Clone, Copy, Eq, PartialEq, Ord, PartialOrd, Hash, Serialize, Deserialize, Debug)]
+pub enum Address {
+    /// Describes an address in the private namespace.
+    Private(XorName),
+    /// Describes an address in the public namespace.
+    Public(XorName),
+}
+
+impl Address {
+    /// Constructs an `Address` given `kind` and `name`.
+    pub fn from_kind(kind: Kind, name: XorName) -> Self {
+        match kind {
+            Kind::Public => Address::Public(name),
+            Kind::Private => Address::Private(name),
+        }
+    }
+
+    /// Returns the kind.
+    pub fn kind(&self) -> Kind {
+        match self {
+            Address::Private(_) => Kind::Private,
+            Address::Public(_) => Kind::Public,
+        }
+    }
+
+    /// Returns the name.
+    pub fn name(&self) -> &XorName {
+        match self {
+            Address::Private(ref name) | Address::Public(ref name) => name,
+        }
+    }
+
+    /// Returns true if public.
+    pub fn is_public(&self) -> bool {
+        self.kind().is_public()
+    }
+
+    /// Returns true if private.
+    pub fn is_private(&self) -> bool {
+        self.kind().is_private()
+    }
+
+    /// Returns the Address serialised and encoded in z-base-32.
+    pub fn encode_to_zbase32(&self) -> String {
+        utils::encode(&self)
+    }
+
+    /// Create from z-base-32 encoded string.
+    pub fn decode_from_zbase32<T: Decodable>(encoded: T) -> Result<Self, Error> {
+        utils::decode(encoded)
+    }
+}
+
+/// Object storing an blob variant.
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Serialize, Deserialize, Debug)]
+pub enum BlobData {
```

Comment:
Looks like we did `ImmutableData -> BlobData` instead of `ImmutableData -> Blob`.

Reply:
Great, some reviews coming in!

This is because of AppendOnlyData. 
It had AppendOnly as the trait, and AppendOnlyData as the enum, corresponding to BlobData here.

At first, AppendOnly trait became Map and Sequence, which made the distinction necessary, calling the enum MapData and SequenceData, which for consistency required BlobData (even though it had no trait Blob).

But then I renamed Map trait to MapBase , and equivalent for Sequence (smacks a bit of OO and inheritance, don't really like to call it base or trait - what is idiomatic here?), which would allow us to ditch "Data" on the enums. 
Will definitely  do that!

Reply:
How about using/reserving `Blob` only for the old `ImmutableData` which contains the datamap of chunks with self-encryption algo, and here we then have `PrivateChunk`/`PublicChunk` instead of the `PrivateBlob`/`PublicBlob`, and `Chunk` instead of `BlobData` ?
So `Blob` is only the user/API facing concept, then we have `Chunks`

Reply:
@bochaco , if I understood you correctly, that would introduce a different pattern than what we do for Map and Sequence.
Have a look at these enums in safe-nd PR #145.

Reply:
Yes, but we could align them also to this other pattern, specially if we consider the other proposal about DataMaps and Chunks, a Blob/Map/Sequence is what we expose as the data types, then each of them are composed of different structures linking chunks. We are otherwise kinda overriding the meaning of a "Blob", is a blob a chunk but also in some other parts a group of chunks with a self-encryption algo to compose what it used to be an ImmutableData file? this is where i see the issue with reusing the name blob for the priv/pub chunk and the actual Blob composed of chunks

Reply:
Oh wow, I had actually misunderstood this data type. It's not the Blob type there, it's just a chunk.. I did not intend for a chunk to be named Blob. 

Yes, this definitely has to be different.

Well, with the other proposal, this also changes. There are no Private or Public chunks, just chunks. Hm, something needs to be done here, but can't really adapt it to the other proposal now.

Reply:
yeah, and but you could even consider a priv/pub chunk enum type to be with/without ref-count respectively

Reply:
Mm, there's a slight break in logic without considering DataMap/Chunk proposal.
But, it's ok. 

From `lib.rs`:

```
pub enum DataType {
    Blob(Chunk),
    Map(Map),
    Sequence(Sequence),
}
```

And it'll be `PrivateChunk`/`PublicChunk`. But I'll do that in the real PR (#147).